### PR TITLE
Refactor ICommunicator API

### PR DESCRIPTION
**Pull request by surfnerd**
- Push (almost) all references to protobuf objects into the RpcCommunicator.
- Simplify the passing around of Agents and Agent Infos.
- Delete all references to the Batcher.
- Simplify the Environment Step by removing all of the reset and message counting logic.

Finishes MLA-27 and MLA-28

### CODE REVIEW

File: UnitySDK/Assets/ML-Agents/Scripts/ICommunicator.cs [hunk 1]
```diff
@@ -46,23 +115,51 @@ exchange call.
...
+        /// <param name="key">Batch Key.</param>
+        /// <param name="agentInfos">Agent info.</param>
+        void PutObservations(string key, Dictionary<Agent,AgentInfo> agentInfos);
+
+        /// <summary>
+        /// Gets the AgentActions based on the batching key.
+        /// </summary>
+        /// <param name="key">A key to identify which actions to get</param>
+        /// <returns></returns>
+        List<Tuple<Agent, AgentAction>> GetActions(string key);
```
Comment (surfnerd) [inline UnitySDK/Assets/ML-Agents/Scripts/ICommunicator.cs]: Not so sure I'm happy with this, but maybe it's better than keeping two lists?

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 2]
```diff
@@ -94,7 +87,18 @@ private void Shutdown()
         /// </summary>
         private void BrainDecideAction()
         {
-            m_BrainBatcher?.SendBrainInfo(name, m_AgentInfos);
+            m_Communicator?.PutObservations(name, m_AgentInfos);
+            var actions = m_Communicator?.GetActions(name);
```
Comment (surfnerd) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: I pulled this logic out of the Communicator to allow brains to fetch actions.

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: This actually changes the behavior to update the actions every time a decision is requested instead of only when every brain has put its observations and a message was sent and received from the communicator

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 3]
```diff
@@ -335,30 +331,38 @@ private void InitializeEnvironment()
...
-            if (newResetParameters != null)
+#if UNITY_EDITOR
+            EditorApplication.isPlaying = false;
+#endif
+            Application.Quit();
+        }
+
+        void OnRLInputReceived(UnityRLInputParameters inputParams)
+        {
+            m_IsInference = inputParams.isTraining;
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Should this be `!inputParams.isTraining`?

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 4]
```diff
@@ -519,33 +523,8 @@ void EnvironmentStep()
                 m_ModeSwitched = false;
             }
 
-            if ((m_IsCommunicatorOn) &&
-                (m_LastCommunicatorMessageNumber != m_BrainBatcher.GetNumberMessageReceived()))
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Can remove m_LastCommunicatorMessageNumber right?

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: yeah, thought I did.  😊

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 3 above)
Comment (chriselion) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Looks like the old Batcher code was `m_IsTraining = firstRlInput.IsTraining`

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: looks like it

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 4 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: fixed

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 5]
```diff
@@ -94,7 +87,18 @@ private void Shutdown()
...
+            if (actions != null)
+                foreach (var tuple in actions)
+                {
+                    tuple.Item1.UpdateVectorAction(tuple.Item2.vectorActions);
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Can you make it clearer what Item1 and Item2 are here? E.g.
``` csharp
var agent = tuple.Item1;
var action = tuple.Item2;
agent.UpdateMemoriesAction(action.memories);
// etc
```

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Yeah, I'm actually wondering if we can just have an UpdateAction(AgentAction action) on agent and remove these other methods.  What do you think about that?

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 6]
```diff
@@ -21,27 +22,19 @@ public abstract class Brain : ScriptableObject
...
-            if (batcher == null)
-            {
-                m_BrainBatcher = null;
-            }
-            else
-            {
-                m_BrainBatcher = batcher;
-                m_BrainBatcher.SubscribeBrain(name);
-            }
+            m_Communicator = communicator ?? null;
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: c# noob question - how is this different from `m_Communicator = communicator;`? m_Communicator will end up null if communicator is null either way, right?

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 7]
```diff
@@ -257,75 +253,70 @@ private void InitializeEnvironment()
             m_OriginalMaximumDeltaTime = Time.maximumDeltaTime;
 
             InitializeAcademy();
-            ICommunicator communicator;
 
             var exposedBrains = broadcastHub.broadcastingBrains.Where(x => x != null).ToList();
             var controlledBrains = broadcastHub.broadcastingBrains.Where(
                 x => x != null && x is LearningBrain && broadcastHub.IsControlled(x));
-            foreach (var brain1 in controlledBrains)
+            var enumerableBrains = controlledBrains as Brain[] ?? controlledBrains.ToArray();
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Why is this necessary ?

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 2 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: How do you allow for the multiple brain scenario? The communicator must still send the messages from all the brains and receive message from all the brains between the PutObservation and the GetAction...

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 8]
```diff
@@ -94,7 +87,18 @@ private void Shutdown()
...
+            if (actions != null)
+                foreach (var tuple in actions)
+                {
+                    tuple.Item1.UpdateVectorAction(tuple.Item2.vectorActions);
+                    tuple.Item1.UpdateMemoriesAction(tuple.Item2.memories);
+                    tuple.Item1.UpdateTextAction(tuple.Item2.textActions);
+                    tuple.Item1.UpdateValueAction(tuple.Item2.value);
+                    tuple.Item1.UpdateCustomAction(tuple.Item2.customAction);
+                }
+
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Why does the brain still get to call DecideAction ? Won't it override the message of the communicator?

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs [hunk 9]
```diff
@@ -4,46 +4,141 @@
...
-            m_CommunicatorParameters = communicatorParameters;
+            m_CommunicatorInitParameters = communicatorInitParameters;
         }
+        
+        #region Initialization
+        
         /// <summary>
-        /// Initialize the communicator by sending the first UnityOutput and receiving the
-        /// first UnityInput. The second UnityInput is stored in the unityInput argument.
+        /// Sends the academy parameters through the Communicator.
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: nit: seems a little scattered on the terminology. "Sends the academy parameters through" but takes a "CommunicatorInitParameters". And it returns "UnityRLInitParameters" but that's used as "var inputParameters" in the Academy setup.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 6 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: yeah, this is gone now.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 7 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: it's not.  I can remove it.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 2 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: It hasn't changed for the multiple brain scenario.  The code was pretty much copy pasted from the batcher into the RpcCommunicator.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 5 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Removed Tuples

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 8 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Where do you think it should go?

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: It's pulling the last message from the communicator every time.  So in theory, it's the same data getting updated in the agent action from the last data received from the communicator.

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs (hunk 9 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: I'll make it more consistent

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: updated.  Let me know if that is better.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 8 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: I think it should either be use he communicator or use he model, not both. I don't understand your comment.

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Funny I don't understand yours either 😂

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 2 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Not sure what the 👀 emoji means.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 10]
```diff
@@ -179,7 +174,7 @@ public abstract class Academy : MonoBehaviour
         bool m_ModeSwitched;
 
         /// Pointer to the batcher currently in use by the Academy.
-        Batcher m_BrainBatcher;
+        ICommunicator m_Communicator;
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: nit: Update "batcher" in comment above.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 11]
```diff
@@ -273,67 +267,59 @@ private void InitializeEnvironment()
...
 
-                academyParameters.EnvironmentParameters =
-                    new CommunicatorObjects.EnvironmentParametersProto();
-                foreach (var key in resetParameters.Keys)
-                {
-                    academyParameters.EnvironmentParameters.FloatParameters.Add(
-                        key, resetParameters[key]
-                    );
-                }
+                m_IsCommunicatorOn = true;
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: Can we get rid of this and just `m_Communicator != null` (or make it a property)?

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 12]
```diff
@@ -345,29 +331,37 @@ private void InitializeEnvironment()
...
 
-        private void UpdateResetParameters()
+        static void OnQuitCommandReceived()
+        {
+#if UNITY_EDITOR
+            EditorApplication.isPlaying = false;
+#endif
+            Application.Quit();
+        }
+        void OnRLInputReceived(UnityRLInputParameters inputParams)
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: nit: whitespace

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 13]
```diff
@@ -18,21 +19,32 @@ public abstract class Brain : ScriptableObject
     {
         [SerializeField] public BrainParameters brainParameters;
 
-        protected Dictionary<Agent, AgentInfo> m_AgentInfos =
-            new Dictionary<Agent, AgentInfo>(1024);
+        protected List<Agent> m_AgentInfos = new List<Agent>(1024);
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: This feels like a bad name now since it no longer has anything to do with AgentInfo

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 14]
```diff
@@ -18,21 +19,32 @@ public abstract class Brain : ScriptableObject
...
+            LazyInitialize();
+        }
+
         /// <summary>
         /// Adds the data of an agent to the current batch so it will be processed in DecideAction.
         /// </summary>
         /// <param name="agent"></param>
-        /// <param name="info"></param>
-        public void SendState(Agent agent, AgentInfo info)
+        public void SendState(Agent agent)
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: This also feels like a bad name since it's not sending anything.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 13 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: fixed

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 14 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: change the name to `Update`

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 12 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: fixed

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 11 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: done

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 10 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: fixed.

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs (hunk 14 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: update is a bad name.

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: Named it `SubscribeAgentForDecision`

File: UnitySDK/Assets/ML-Agents/Scripts/Brain.cs [hunk 15]
```diff
@@ -18,21 +19,32 @@ public abstract class Brain : ScriptableObject
...
+        /// Updates the data of an agent to the current batch so it will be processed in DecideAction.
         /// </summary>
         /// <param name="agent"></param>
-        /// <param name="info"></param>
-        public void SendState(Agent agent, AgentInfo info)
+        public void Update(Agent agent)
         {
             LazyInitialize();
-            m_AgentInfos[agent] = info;
+            m_Agents.Add(agent);
```
Comment (chriselion) [inline UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: (summarizing slack thread)
This feels odd because we always append to the agent list here, and rely on the implementations to clear it. It's slightly worse than before, because if something prevents the clearing (i.e. an exception due to bad configuration), the list grows unbounded now, instead of overwriting when it was a dictionary.

Not sure if we should leave as is, convert to a Set, or change how the clearing happens.

@vincentpierre Ideas?

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Brain.cs]: changed here 588466436341384aff58b55ff79cd811719f522f

File: UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs [hunk 16]
```diff
@@ -50,16 +49,6 @@ public class LearningBrain : Brain
         private IReadOnlyList<TensorProxy> m_InferenceInputs;
         private IReadOnlyList<TensorProxy> m_InferenceOutputs;
 
-        /// <summary>
-        /// When Called, the brain will be controlled externally. It will not use the
-        /// model to decide on actions.
-        /// </summary>
-        public void SetBatcher(Batcher batcher)
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs]: In my previous PR, I removed the capacity of Non-Learning Brains to have a communicator. It seems you are re-introducing it here by giving the abstract Brain the m_Communicator. Please make this a LearningBrain thing only.

Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs]: done.

**Comment by surfnerd** [comment]
> It seems walljump is broken
> The editor does no longer exit playmode when ctrl-C on the terminal
@vincentpierre 
That could be related to #2620.  I didn't change any of the shutdown code related to this.  Is the wall jump issue separate from the ctrl-C issue?

**Comment by vincentpierre** [comment]
Shutdown works on develop but not on this branch
walljump is a different issue that I think is due to multi-brain training

**Comment by surfnerd** [comment]
@vincentpierre fixes are in the latest commit aae9487d.  Let me know if those work for you.

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs [hunk 17]
```diff
@@ -108,11 +367,13 @@ public UnityInputProto Exchange(UnityOutputProto unityOutput)
...
-                {
-                    m_IsOpen = false;
-                    return null;
-                }
+
+                m_IsOpen = false;
+                // Not sure if the quit command is actually sent when a
+                // non 200 message is received.  Notify that we are indeed
+                // quitting.
+                QuitCommandReceived?.Invoke();
```
[the remark this answers is not in the document]
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: This previously was a "hack".  It would set the m_Command to Quit.

[the remark this answers is not in the document]
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: https://github.com/Unity-Technologies/ml-agents/blob/866ed5c4c4762d0d1f3b1e58c2b8e486b3c036ec/ml-agents-envs/mlagents/envs/rpc_communicator.py#L118

there is no quit command sent.

**Comment by surfnerd** [comment]
Fixed the reset issue dfd4056

File: test.py [hunk 18]
```diff
@@ -0,0 +1,19 @@
+from mlagents.envs.environment import UnityEnvironment
```
Comment (vincentpierre) [inline test.py]: Delete

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs [hunk 19]
```diff
@@ -12,38 +15,130 @@ namespace MLAgents
...
+                throw new UnityAgentsException(exceptionMessage);
+            }
+
+            UpdateEnvironmentWithInput(input.RlInput);
+            return initializationInput.RlInitializationInput.ToUnityRLInitParameters();
+        }
+
+        void UpdateEnvironmentWithInput(UnityRLInputProto rlInput)
+        {
+            Debug.Log("rlInput.Command: " + rlInput.Command);
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: delete

File: test.py (hunk 18 above)
Comment (surfnerd) [reply test.py]: gone

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs (hunk 19 above)
Comment (surfnerd) [reply UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: gone

File: UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs [hunk 20]
```diff
@@ -108,11 +367,13 @@ public UnityInputProto Exchange(UnityOutputProto unityOutput)
...
-                }
+
+                m_IsOpen = false;
+                // Not sure if the quit command is actually sent when a
+                // non 200 message is received.  Notify that we are indeed
+                // quitting.
+                QuitCommandReceived?.Invoke();
+                return message.UnityInput;
             }
             catch
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs]: If there is an error, the quitCommand should be thrown.

**Comment by surfnerd** [comment]
Thank you @vincentpierre !
